tests/base.py

```diff
@@ -211,52 +221,128 @@ def run_and_verify_sync(self, conn_id):
+            )
+            if is_field_metadata and inclusion_automatic_or_selected:
+                selected_fields.add(field['breadcrumb'][1])
+        return selected_fields
+
+    @staticmethod
+    def select_all_streams_and_fields(conn_id, catalogs, select_all_fields: bool = True):
+        """Select all streams and all fields within streams"""
+        for catalog in catalogs:
+            schema = menagerie.get_annotated_schema(conn_id, catalog['stream_id'])
 
-    def timedelta_formatted(self, dtime, days=0, str_format="%Y-%m-%dT00:00:00Z"):
-        date_stripped = dt.strptime(dtime, str_format)
-        return_date = date_stripped + timedelta(days=days)
+            non_selected_properties = []
+            if not select_all_fields:
+                # Get a list of all properties so that none are selected
+                non_selected_properties = schema.get('annotated-schema', {}).get(
+                    'properties', {}).keys()
 
-        return dt.strftime(return_date, str_format)
+            connections.select_catalog_and_fields_via_metadata(
+                conn_id, catalog, schema, [], non_selected_properties)
 
     ################################
     #  Tap Specific Test Actions   #
     ################################
+
+    def dt_to_ts(self, dtime, format):
+        """Convert datetime with a format to timestamp"""
+        date_stripped = int(time.mktime(dt.strptime(dtime, format).timetuple()))
+        return date_stripped
+
+    def calculated_states_by_stream(self, current_state):
+        """
+        Look at the bookmarks from a previous sync and set a new bookmark
+        value based off timedelta expectations. This ensures the subsequent sync will replicate
+        at least 1 record but, fewer records than the previous sync.
+
+        Sufficient test data is required for this test to cover a given stream.
+        An incremental replication stream must have at least two records with
+        replication keys that differ by some time span.
+
+        If the test data is changed in the future this may break expectations for this test.
+        """
+        timedelta_by_stream = {stream: [0, 12, 0]  # {stream_name: [days, hours, minutes], ...}
+                               for stream in current_state['bookmarks'].keys()}
+
+        stream_to_calculated_state = {
+            stream: "" for stream in current_state['bookmarks'].keys()}
+        for stream, state in current_state['bookmarks'].items():
+            state_key, state_value = list(state.keys())[0], list(state.values())[0]
+            state_as_datetime = dateutil.parser.parse(state_value)
+
+            days, hours, minutes = timedelta_by_stream[stream]
+            calculated_state_as_datetime = state_as_datetime - timedelta(days=days, hours=hours, minutes=minutes)
+
+            state_format = self.BOOKMARK_FORMAT
+            calculated_state_formatted = datetime.datetime.strftime(calculated_state_as_datetime, state_format)
```

Comment:
Instead of `datetime.datetime` use `dt`.

Reply:
Replaced datetime.datetime with dt.